File: cmd/configure.go
package cmd

import (
	"encoding/json"
	"fmt"

	"github.com/spf13/cobra"
)

var configureCmd = &cobra.Command{
	Use:   "configure",
	Short: "Configure session settings",
	RunE:  runConfigure,
}

func init() {
	configureCmd.GroupID = "agent"
	rootCmd.AddCommand(configureCmd)
	configureCmd.Flags().String("session-id", "", "session ID (auto-detected if omitted)")
	configureCmd.Flags().String("inbox-mode", "", "inbox mode (manual, on-submit, auto)")
	configureCmd.Flags().Int("auto-poll-interval", 0, "auto-poll interval in seconds (for auto mode)")
	configureCmd.Flags().String("role", "", "session role (handler, or empty to clear)")
	configureCmd.Flags().String("get", "", "get a specific setting value (inbox-mode, auto-poll-interval, role)")
}

func runConfigure(cmd *cobra.Command, args []string) error {
	sessionID, err := resolveSessionID(cmd)
	if err != nil {
		return fmt.Errorf("could not determine session: %w", err)
	}

	getFlag, _ := cmd.Flags().GetString("get")

	if getFlag != "" {
		d, err := openReadOnlyDB()
		if err != nil {
			return fmt.Errorf("failed to open database: %w", err)
		}
		defer d.Close()

		session, err := d.GetSession(sessionID)
		if err != nil {
			return fmt.Errorf("failed to get session: %w", err)
		}

		switch getFlag {
		case "inbox-mode":
			fmt.Println(session.InboxMode)
		case "auto-poll-interval", "auto_poll_interval":
			if session.AutoPollInterval != nil {
				fmt.Println(*session.AutoPollInterval)
			} else {
				fmt.Println("null")
			}
		case "role":
			fmt.Println(session.Role)
		default:
			return fmt.Errorf("unknown setting: %s (valid: inbox-mode, auto-poll-interval, role)", getFlag)
		}
		return nil
	}

	inboxMode, _ := cmd.Flags().GetString("inbox-mode")
	autoPollInterval, _ := cmd.Flags().GetInt("auto-poll-interval")
	roleFlag, _ := cmd.Flags().GetString("role")

	if inboxMode == "" && autoPollInterval == 0 && !cmd.Flags().Changed("role") {
		return fmt.Errorf("at least one of --inbox-mode, --auto-poll-interval, or --role must be provided")
	}

	d, err := openDB()
	if err != nil {
		return fmt.Errorf("failed to open database: %w", err)
	}
	defer d.Close()

	session, err := d.GetSession(sessionID)
	if err != nil {
		return fmt.Errorf("failed to get session: %w", err)
	}

	finalInboxMode := session.InboxMode
	if inboxMode != "" {
		finalInboxMode = inboxMode
	}

	var finalAutoPoll *int
	if autoPollInterval > 0 {
		finalAutoPoll = &autoPollInterval
	} else {
		finalAutoPoll = session.AutoPollInterval
	}

	var finalRole *string
	if cmd.Flags().Changed("role") {
		if roleFlag == "" {
			// Explicit empty string means clear the role
			finalRole = new(string)
		} else {
			finalRole = &roleFlag
		}
	}

	if err := d.ConfigureSession(sessionID, finalInboxMode, finalAutoPoll, finalRole); err != nil {
		return fmt.Errorf("failed to configure session: %w", err)
	}

	// Re-fetch session to get final state
	session, err = d.GetSession(sessionID)
	if err != nil {
		return fmt.Errorf("failed to get session: %w", err)
	}

	if jsonOutput {
		output := map[string]interface{}{
			"session_id":         sessionID,
			"inbox_mode":         session.InboxMode,
			"auto_poll_interval": session.AutoPollInterval,
			"role":               session.Role,
		}
		data, err := json.MarshalIndent(output, "", "  ")
		if err != nil {
			return err
		}
		fmt.Println(string(data))
	} else {
		fmt.Printf("✓ Configured session %s\n", sessionID)
		fmt.Printf("  Inbox mode: %s\n", session.InboxMode)
		if session.AutoPollInterval != nil {
			fmt.Printf("  Auto-poll interval: %d seconds\n", *session.AutoPollInterval)
		}
		if session.Role != "" {
			fmt.Printf("  Role: %s\n", session.Role)
		}
	}

	return nil
}

File: src/Handler/Cmd/Configure.hs
-- | Port of cmd/configure.go: configure session settings.
module Handler.Cmd.Configure (configureCommand) where

import Control.Monad (when)
import Data.Aeson (object, (.=))
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import Options.Applicative

import Handler.Cli.Common
import Handler.Db (close)
import Handler.Db.Cursors (catchUpHumanCursor, clearHumanCursor)
import Handler.Db.Sessions (Session(..), configureSession, getSession)
import Handler.Util (textShow)

data ConfigureOpts = ConfigureOpts
  { cSessionId :: Maybe Text
  , cInboxMode :: Text
  , cInterval  :: Int
  , cRole      :: Maybe Text
  , cGet       :: Text
  }

configureCommand :: Mod CommandFields NamedCommand
configureCommand = mkCommand "configure" "Configure session settings" (runConfigure <$> opts)
  where
    opts = ConfigureOpts
      <$> sessionIdOption
      <*> strOption (long "inbox-mode" <> value "" <> help "inbox mode (manual, on-submit, auto)")
      <*> option auto (long "auto-poll-interval" <> value 0
            <> help "auto-poll interval in seconds (for auto mode)")
      <*> optional (strOption (long "role" <> help "session role (handler, or empty to clear)"))
      <*> strOption (long "get" <> value ""
            <> help "get a specific setting value (inbox-mode, auto-poll-interval, role)")

runConfigure :: ConfigureOpts -> Ctx -> IO ()
runConfigure o ctx = do
  sessionId <- resolveSessionIdOpt o.cSessionId

  if o.cGet /= ""
    then do
      db <- openReadOnlyDb
      msession <- getSession db sessionId
      case msession of
        Nothing -> dieText ("session " <> T.pack (show sessionId) <> " not found")
        Just session -> case o.cGet of
          "inbox-mode" -> putTextLn session.inboxMode
          g | g == "auto-poll-interval" || g == "auto_poll_interval" ->
            putTextLn (maybe "null" textShow session.autoPollInterval)
          "role" -> putTextLn session.role
          other -> dieText ("unknown setting: " <> other
                            <> " (valid: inbox-mode, auto-poll-interval, role)")
      close db
    else do
      when (o.cInboxMode == "" && o.cInterval == 0 && o.cRole == Nothing) $
        dieText "at least one of --inbox-mode, --auto-poll-interval, or --role must be provided"

      db <- openDb
      msession <- getSession db sessionId
      case msession of
        Nothing -> dieText ("session " <> T.pack (show sessionId) <> " not found")
        Just session -> do
          let finalInboxMode = if o.cInboxMode == "" then session.inboxMode else o.cInboxMode
              finalAutoPoll = if o.cInterval > 0 then Just o.cInterval else session.autoPollInterval
          configureSession db sessionId finalInboxMode finalAutoPoll o.cRole

          -- Initialize human cursor when entering auto mode
          when (o.cInboxMode == "auto" && session.inboxMode /= "auto") $
            catchUpHumanCursor db sessionId
          -- Clear human cursor when leaving auto mode
          when (o.cInboxMode /= "" && o.cInboxMode /= "auto" && session.inboxMode == "auto") $
            clearHumanCursor db sessionId

          -- Re-fetch session to get final state
          mfinal <- getSession db sessionId
          case mfinal of
            Nothing -> dieText "failed to get session"
            Just final ->
              if ctx.jsonOutput
                then printJson $ object
                  [ "session_id" .= sessionId
                  , "inbox_mode" .= final.inboxMode
                  , "auto_poll_interval" .= final.autoPollInterval
                  , "role" .= final.role
                  ]
                else do
                  printfT ["✓ Configured session ", sessionId]
                  printfT ["  Inbox mode: ", final.inboxMode]
                  case final.autoPollInterval of
                    Just interval -> printfT ["  Auto-poll interval: ", textShow interval, " seconds"]
                    Nothing -> pure ()
                  when (final.role /= "") $
                    printfT ["  Role: ", final.role]
      close db
